### PR TITLE
[CPROD-914] BE:: convert IS20 to use principal and subaccounts

### PR DESCRIPTION
**Pull request by itsyaasir**

### CODE REVIEW

File: spec/IS20.md [hunk 1]
```diff
@@ -97,19 +127,19 @@ type TransactionStatus = variant {
 };
 
 type TxRecord = record {
-  to : principal;
-  fee : nat;
+  to : AccountIdentifier;
+  fee : Tokens128;
   status : TransactionStatus;
-  from : principal;
+  from_details : opt(Principal, Subaccount);
```
Comment (blutooth) [inline spec/IS20.md:133]: from: (principal, subaccount)

File: spec/IS20.md [hunk 2]
```diff
@@ -97,19 +127,19 @@ type TransactionStatus = variant {
...
   operation : Operation;
   timestamp : int;
-  caller : opt principal;
-  index : nat;
-  amount : nat;
+  caller : opt AccountIdentifier;
```
Comment (blutooth) [inline spec/IS20.md:136]: caller: principal

File: src/token/api/src/canister/erc20_transactions.rs [hunk 3]
```diff
@@ -1,59 +1,26 @@
...
 use crate::canister::is20_auction::auction_principal;
-use crate::principal::{CheckedPrincipal, Owner, SenderRecipient, TestNet, WithRecipient};
+use crate::principal::{CheckedPrincipal, Owner, TestNet, WithRecipient};
 use crate::state::{Balances, CanisterState};
-use crate::types::{TxError, TxReceipt};
+use crate::types::{Account, Subaccount, TxError, TxReceipt};
 
 use super::TokenCanisterAPI;
 
 pub fn transfer(
```
Comment (blutooth) [inline src/token/api/src/canister/erc20_transactions.rs:12]: icrc_transfer

File: src/token/api/src/canister.rs [hunk 4]
```diff
@@ -392,12 +394,12 @@ pub trait TokenCanisterAPI: Canister + Sized {
     #[query(trait = true)]
     fn getTransactions(
         &self,
-        who: Option<Principal>,
+        who: Option<Account>,
```
Comment (blutooth) [inline src/token/api/src/canister.rs:397]: get all accounts from a principal

File: src/token/api/src/canister.rs [hunk 5]
```diff
@@ -406,8 +408,9 @@ pub trait TokenCanisterAPI: Canister + Sized {
 
     /// Returns the total number of transactions related to the user `who`.
     #[query(trait = true)]
-    fn getUserTransactionCount(&self, who: Principal) -> usize {
-        self.state().borrow().ledger.get_len_user_history(who)
+    fn getUserTransactionCount(&self, who: Principal, who_subaccount: Option<Subaccount>) -> usize {
```
Comment (blutooth) [inline src/token/api/src/canister.rs:411]: same, just from a principal

File: spec/IS20.md [hunk 6]
```diff
@@ -51,24 +70,35 @@ type Metadata = {
...
    symbol : text; // token symbol
    decimals : nat8; // token decimal
-   totalSupply : nat; // token total supply
+   totalSupply : Tokens128; // token total supply
    owner : principal; // token owner
    logo : text; // base64 encoded logo or logo url
-   fee : nat; // fee for update calls
+   fee : Tokens128; // fee for update calls
 }
 ```
```
Comment (blutooth) [inline spec/IS20.md:79]: mint(to: AccountIdentifier, value){
account_balances[accID] += value 
}

transfer_from_any(x) {
// while loop 
}

fn transfer(){
//check that the accID formed from caller() has balance
}

fn transfer_toaccountid(accid val) {
mintees[accID] = value 
// store principal in the tx
}

fn

File: src/token/api/src/canister/is20_transactions.rs [hunk 7]
```diff
@@ -80,12 +85,14 @@ pub fn batch_transfer(
...
         .ok_or(TxError::AmountOverflow)?;
 
-    if balances.balance_of(&from) < (total_value + total_fee).ok_or(TxError::AmountOverflow)? {
+    if balances.balance_of(from) < (total_value + total_fee).ok_or(TxError::AmountOverflow)? {
         return Err(TxError::InsufficientBalance);
     }
 
     {
-        for (to, value) in transfers.clone() {
+        for x in transfers.clone() {
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:93]: easy enough to read today, but perhaps not in the future.
change `x` to `transfer`
```suggestion
        for x in &transfers {
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 8]
```diff
@@ -35,35 +39,36 @@ pub fn transfer_include_fee(
...
         (amount - fee).expect("amount > fee is checked above"),
     )
     .expect("never fails due to checks above");
 
-    let id = ledger.transfer(caller.inner(), caller.recipient(), amount, fee);
+    let id = ledger.transfer(from, to, amount, fee);
     Ok(id)
 }
 
 pub fn batch_transfer(
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:59]: Change the state deconstruction to: 
```
    let CanisterState {
        ref mut balances,
        ref mut ledger,
        ref bidding_state,
        ref stats,
        ..
    } = &mut *state;
```
so we can call `let id = ledger.batch_transfer(from, transfers, fee);` on the penultimate line of this function

File: src/token/api/src/ledger.rs [hunk 9]
```diff
@@ -41,7 +43,11 @@ impl Ledger {
             .history
             .iter()
             .rev()
-            .filter(|tx| who.map_or(true, |c| c == tx.from || c == tx.to || Some(c) == tx.caller))
+            .filter(|tx| {
+                who.map_or(true, |c| {
```
Comment (hagsteel) [inline src/token/api/src/ledger.rs:47]: add a `contains` method `TxRectord` and then update the filter to `filter(|&tx| who.map_or(true, |c| tx.contains(c)))`

File: src/token/api/src/state.rs [hunk 10]
```diff
@@ -31,31 +36,11 @@ impl CanisterState {
...
-            .unwrap_or(0)
-    }
-
-    pub fn user_approvals(&self, who: Principal) -> Vec<(Principal, Tokens128)> {
-        match self.allowances.get(&who) {
-            Some(allow) => Vec::from_iter(allow.clone().into_iter()),
-            None => Vec::new(),
-        }
+    pub fn claim_amount(&self, account: AccountIdentifier) -> Tokens128 {
+        *self.claims.get(&account).unwrap_or(&Tokens128::ZERO)
```
Comment (hagsteel) [inline src/token/api/src/state.rs:40]:
```suggestion
        self.claims.get(&account).copied().unwrap_or(Tokens128::ZERO)
```

File: src/token/api/src/state.rs [hunk 11]
```diff
@@ -65,24 +50,71 @@ impl Versioned for CanisterState {
...
+            .get(&account.account)
+            .and_then(|subaccounts| subaccounts.get(&account.subaccount))
+            .unwrap_or(&Tokens128::default())
+    }
+
+    pub fn get_holders(
+        &self,
+        start: usize,
+        limit: usize,
+    ) -> Vec<((Principal, Subaccount), Tokens128)> {
```
Comment (hagsteel) [inline src/token/api/src/state.rs:80]: should probably return `Vec<Account, Tokens128>` instead

File: src/token/api/src/state.rs [hunk 12]
```diff
@@ -65,24 +50,71 @@ impl Versioned for CanisterState {
...
+            .iter()
+            .flat_map(|(principal, subaccounts)| {
+                subaccounts
+                    .iter()
+                    .map(|(subaccount, token)| ((*principal, *subaccount), *token))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+        holders.sort_by(|a, b| b.1.cmp(&a.1));
+        holders[start..].iter().take(limit).cloned().collect()
```
Comment (hagsteel) [inline src/token/api/src/state.rs:92]:
```suggestion
            .skip(start)
            .take(limit)
            .collect::<Vec<_>>();
        holders.sort_by(|a, b| b.1.cmp(&a.1));
        holders
```

File: src/token/api/src/state.rs [hunk 13]
```diff
@@ -65,24 +50,71 @@ impl Versioned for CanisterState {
...
     }
 
-    pub fn get_holders(&self, start: usize, limit: usize) -> Vec<(Principal, Tokens128)> {
-        let mut balance = self.0.iter().map(|(&k, v)| (k, *v)).collect::<Vec<_>>();
+    pub fn get_balances(&self, who: &Principal) -> HashMap<Subaccount, Tokens128> {
+        self.0.get(who).cloned().unwrap_or_default()
+    }
```
Comment (hagsteel) [inline src/token/api/src/state.rs:97]: Remove this function

File: src/token/api/src/types/account.rs [hunk 14]
```diff
@@ -0,0 +1,34 @@
+use std::fmt::{Display, Formatter};
```
Comment (hagsteel) [inline src/token/api/src/types/account.rs:1]: move this to the tokens project instead, we need to get rid of types.rs

**Review by hagsteel — DISMISSED** [review]
You know what to do !!!

**Review by hagsteel — APPROVED** [review]

**Review by awkure — APPROVED** [review]

File: spec/IS20.md [hunk 15]
```diff
@@ -51,24 +70,33 @@ type Metadata = {
...
-  Unauthorized;
-  AmountTooSmall,
-  NotificationFailed;
-  AlreadyNotified;
-  TransactionDoesNotExist;
+    InsufficientBalance,
+    Unauthorized,
+    AmountTooSmall,
+    FeeExceededLimit,
+    ApproveSucceededButNotifyFailed,
```
Comment (blutooth) [inline spec/IS20.md:88]: del

File: spec/IS20.md [hunk 16]
```diff
@@ -119,63 +143,60 @@ The update calls described in this section might choose to charge `fee` amount o
...
+
+This method returns a `TxReceipt` which contains the transaction index or an error message. The balance of the caller is
+reduced by `value + fee` amount.
 
 To protect the caller from unexpected fee amount change, the optional `fee_limit` parameter can be given. If the
 fee to be applied is larger than this value, the transaction will fail with `TxError::FeeExceededLimit` error.
 
 ```
-update transfer(to: Principal, value: nat, fee_limit: opt nat) : TxReceipt
+update icrc1_tranfer(from_subaccount: opt Subaccount, to: Principal, to_subaccount: opt Subaccount, amount: Tokens128, fee_limit: opt Tokens128) -> TxReceipt;
```
Comment (blutooth) [inline spec/IS20.md:163]: arguments must match type  of icrc1 inc;uding nat

File: src/token/api/src/canister.rs [hunk 17]
```diff
@@ -133,28 +136,14 @@ pub trait TokenCanisterAPI: Canister + Sized {
...
-
-    #[query(trait = true)]
-    fn balanceOf(&self, holder: Principal) -> Tokens128 {
-        self.state().borrow().balances.balance_of(&holder)
-    }
-
-    #[query(trait = true)]
-    fn allowance(&self, owner: Principal, spender: Principal) -> Tokens128 {
-        self.state().borrow().allowance(owner, spender)
+    fn balanceOf(&self, holder: Principal, holder_subaccount: Option<Subaccount>) -> Tokens128 {
```
Comment (blutooth) [inline src/token/api/src/canister.rs:144]: same as ICRC_1 spec?

File: src/token/api/src/canister.rs [hunk 18]
```diff
@@ -212,28 +201,21 @@ pub trait TokenCanisterAPI: Canister + Sized {
...
-    #[update(trait = true)]
-    fn approve(&self, spender: Principal, amount: Tokens128) -> TxReceipt {
-        let caller = CheckedPrincipal::with_recipient(spender)?;
-        approve(self, caller, amount)
-    }
-
     /********************** TRANSFERS ***********************/
     #[cfg_attr(feature = "transfer", update(trait = true))]
-    fn transfer(
+    fn icrc1_transfer(
```
Comment (blutooth) [inline src/token/api/src/canister.rs:206]: icrc_! spec?

File: src/token/api/src/canister.rs [hunk 19]
```diff
@@ -276,21 +286,42 @@ pub trait TokenCanisterAPI: Canister + Sized {
     /// If `from` is Some(_) but method called not by owner, `TxError::Unauthorized` will be returned.
     /// If owner calls this method and `from` is Some(who), then who's tokens will be burned.
     #[cfg_attr(feature = "mint_burn", update(trait = true))]
-    fn burn(&self, from: Option<Principal>, amount: Tokens128) -> TxReceipt {
+    fn icrc1_burn(
```
Comment (blutooth) [inline src/token/api/src/canister.rs:289]: same as ircrc1 _spec?

File: src/token/api/src/canister/is20_notify.rs [hunk 20]
```diff
@@ -111,31 +98,6 @@ mod tests {
...
-
-        let canister = test_canister();
-
-        canister.approveAndNotify(bob(), AMOUNT).await.unwrap();
-        assert!(is_notified_clone.load(Ordering::Relaxed));
-        assert_eq!(counter_copy.load(Ordering::Relaxed), 1);
-    }
-
     #[tokio::test]
     async fn notify_non_existing() {
```
Comment (blutooth) [inline src/token/api/src/canister/is20_notify.rs:102]: delete notify

File: src/token/api/src/state.rs [hunk 21]
```diff
@@ -31,31 +35,14 @@ impl CanisterState {
...
-            .reduce(|accum, v| accum + v)
-            .unwrap_or(0)
-    }
-
-    pub fn user_approvals(&self, who: Principal) -> Vec<(Principal, Tokens128)> {
-        match self.allowances.get(&who) {
-            Some(allow) => Vec::from_iter(allow.clone().into_iter()),
-            None => Vec::new(),
-        }
+    pub fn claim_amount(&self, account: AccountIdentifier) -> Tokens128 {
```
Comment (blutooth) [inline src/token/api/src/state.rs:38]: public method get_claim(opt subaccount) -> amount

**Pull request closed without merging**